File: Compiler/Parser/Expr.py
import Compiler.Syntax.Expr as Expr
import Compiler.Parser.Function as Function
import Compiler.Parser.Pattern as Pattern
import Compiler.Token as Token

def parseFunctionArgs(parser):
    parser.expect(Token.LeftParen())
    args = []
    while not parser.peek(Token.RightParen()):
        args.append(parseExpr(parser))
        if parser.peek(Token.RightParen()):
            break
        else:
            parser.expect(Token.Comma())
    parser.expect(Token.RightParen())
    return args

def parseFunctionCall(parser):
    receiver = parsePrimary(parser)
    while True:
        if parser.peek(Token.LeftParen()):
            call = Expr.FunctionCall()
            call.id = receiver
            call.args = parseFunctionArgs(parser)
            receiver = call
        elif parser.peek(Token.Dot()):
            parser.expect(Token.Dot())
            name = parser.parseName()
            if parser.peek(Token.LeftParen()):
                args = parseFunctionArgs(parser)
                m = Expr.MemberCall()
                m.receiver = receiver
                m.name = name
                m.args = args
                receiver = m
            else:
                m = Expr.MemberAccess()
                m.receiver = receiver
                m.name = name
                receiver = m
        else:
            break
    return receiver

def parseIf(parser):
    parser.expect("if")
    if_expr = Expr.If()
    if_expr.cond = parseExpr(parser)
    if_expr.true_branch = Function.parseBlock(parser)
    if parser.peek("else"):
        parser.expect("else")
        if_expr.false_branch = Function.parseBlock(parser)
    return if_expr

def parseMatch(parser):
    parser.expect("match")
    match_expr = Expr.Match()
    match_expr.body = parseExpr(parser)
    parser.expect(Token.LeftCurly())
    while True:
        branch = Expr.MatchBranch()
        match_expr.branches.append(branch)
        branch.pattern = Pattern.parsePattern(parser)
        parser.expect(Token.RightArrow())
        branch.body = parseExpr(parser)
        if parser.peek(Token.Comma()):
            parser.expect(Token.Comma())
        if parser.peek(Token.RightCurly()):
            break
    parser.expect(Token.RightCurly())
    return match_expr

def parseLoop(parser):
    parser.expect("loop")
    var = parser.parseName()
    parser.expect("equal")
    init = parser.parseExpr()
    body = parser.parseBlock()
    loop_expr = Expr.Loop()
    loop_expr.var = var
    loop_expr.init = init
    loop_expr.body = body
    return loop_expr

def parseForLoop(parser):
    parser.expect("for")
    pattern = Pattern.parsePattern(parser)
    parser.expect("in")
    init = parseExpr(parser)
    body = Function.parseBlock(parser)
    loop_expr = Expr.ForLoop()
    loop_expr.pattern = pattern
    loop_expr.init = init
    loop_expr.body = body
    return loop_expr

def parsePrimary(parser):
    if parser.peek("string"):
        l = Expr.StringLiteral()
        l.value = parser.tokens[parser.index].value
        parser.step()
        return l
    if parser.peek("typeid"):
        name = parser.parseQualifiedName()
        r = Expr.TypeRef()
        r.name = name
        return r
    elif parser.peek("varid"):
        name = parser.parseName()
        r = Expr.VarRef()
        r.name = name
        return r
    elif parser.peek(Token.LeftCurly()):
        e = Function.parseBlock(parser)
        return e
    elif parser.peek("break"):
        parser.expect("break")
        e = Expr.Break()
        e.arg = parseExpr(parser)
        return e
    elif parser.peek("continue"):
        parser.expect("continue")
        e = Expr.Continue()
        if not parser.peek(Token.Semicolon()) and not parser.peek(Token.Comma()):
            e.arg = parseExpr(parser)
        return e
    elif parser.peek("return"):
        parser.expect("return")
        e = Expr.Return()
        e.arg = parseExpr(parser)
        return e
    elif parser.peek("loop"):
        return parseLoop(parser)
    elif parser.peek("match"):
        return parseMatch(parser)
    elif parser.peek("for"):
        return parseForLoop(parser)
    elif parser.peek("if"):
        return parser.parseIf()
    elif parser.peek(Token.LeftParen()):
        parser.expect(Token.LeftParen())
        items = []
        if not parser.peek(Token.RightParen()):
            while True:
                item = parseExpr(parser)
                items.append(item)
                if parser.peek(Token.Comma()):
                    parser.expect(Token.Comma())
                else:
                    break
        parser.expect(Token.RightParen())
        if len(items) == 1:
            return items[0]
        e = Expr.Tuple()
        e.args = items
        return e
    elif parser.peek("true"):
        parser.expect("true")
        e = Expr.BoolLiteral()
        e.value = True
        return e
    elif parser.peek("false"):
        parser.expect("false")
        e = Expr.BoolLiteral()
        e.value = False
        return e
    else:
        parser.error("expected expr, found %s" % parser.tokens[parser.index].type)

def opsTable():
    ops_table = [[Token.And(), Token.Or()],
                  [Token.DoubleEqual(), Token.NotEqual()],
                  [Token.LessThan(),
                   Token.GreaterThan(),
                   Token.LessThanOrEqual(),
                   Token.GreaterThanOrEqual()],
                  [Token.Plus(), Token.Minus()],
                  [Token.Mul(), Token.Div()]]
    return ops_table

def callNext(parser, index):
    if index + 1 < len(opsTable()):
        return parseBinary(parser, index + 1)
    else:
        return parseFunctionCall(parser)

def parseBinary(parser, index):
    expr = callNext(parser, index)
    while True:
        ops = opsTable()[index]
        matchingOp = None
        for op in ops:
            if parser.peek(op):
                matchingOp = op
                break
        if matchingOp is not None:
            parser.expect(op)
            rhs = callNext(parser, index)
            e = Expr.BinaryOp()
            e.op = op
            e.rhs = rhs
            e.lhs = expr
            expr = e
        else:
            break
    return expr

def parseExpr(parser):
    return parseBinary(parser, 0)

File: sikors/src/siko/parser/Expr.rs
use crate::siko::syntax::{
    Expr::{BinaryOp, Branch, Expr},
    Statement::{Block, Statement, StatementKind},
};

use super::{
    Parser::*,
    Pattern::PatternParser,
    Token::{ArrowKind, BracketKind, KeywordKind, MiscKind, OperatorKind, Token, TokenKind},
};

pub trait ExprParser {
    fn parseBlock(&mut self) -> Block;
    fn parseStatement(&mut self) -> (StatementKind, SemicolonRequirement);
    fn parseIf(&mut self) -> Expr;
    fn parseFor(&mut self) -> Expr;
    fn parseMatch(&mut self) -> Expr;
    fn parseFieldAccessOrCall(&mut self) -> Expr;
    fn parseBinaryOp(&mut self, index: usize) -> Expr;
    fn parseExpr(&mut self) -> Expr;
    fn parsePrimary(&mut self) -> Expr;
    fn callNext(&mut self, index: usize) -> Expr;
}

pub enum SemicolonRequirement {
    Optional,
    TrailingOptional,
    Required,
}

impl ExprParser for Parser {
    fn parseBlock(&mut self) -> Block {
        let mut statements = Vec::new();
        self.expect(TokenKind::LeftBracket(BracketKind::Curly));
        while !self.check(TokenKind::RightBracket(BracketKind::Curly)) {
            let (statementKind, requirement) = self.parseStatement();
            let trailing = self.check(TokenKind::RightBracket(BracketKind::Curly));
            let mut hasSemicolon = false;
            match requirement {
                SemicolonRequirement::Optional => {
                    if self.check(TokenKind::Misc(MiscKind::Semicolon)) {
                        hasSemicolon = true;
                        self.expect(TokenKind::Misc(MiscKind::Semicolon));
                    }
                }
                SemicolonRequirement::TrailingOptional => {
                    if trailing {
                        if self.check(TokenKind::Misc(MiscKind::Semicolon)) {
                            hasSemicolon = true;
                            self.expect(TokenKind::Misc(MiscKind::Semicolon));
                        }
                    } else {
                        hasSemicolon = true;
                        self.expect(TokenKind::Misc(MiscKind::Semicolon));
                    }
                }
                SemicolonRequirement::Required => {
                    hasSemicolon = true;
                    self.expect(TokenKind::Misc(MiscKind::Semicolon));
                }
            }

            statements.push(Statement {
                kind: statementKind,
                hasSemicolon,
            });
        }
        self.expect(TokenKind::RightBracket(BracketKind::Curly));
        Block { statements }
    }

    fn parseIf(&mut self) -> Expr {
        self.expect(TokenKind::Keyword(KeywordKind::If));
        let cond = self.parseExpr();
        self.expect(TokenKind::LeftBracket(BracketKind::Curly));
        self.undo();
        let trueBranch = self.parseExpr();
        let falseBranch = if self.check(TokenKind::Keyword(KeywordKind::Else)) {
            self.expect(TokenKind::Keyword(KeywordKind::Else));
            self.expect(TokenKind::LeftBracket(BracketKind::Curly));
            self.undo();
            let falseBranch = self.parseExpr();
            Some(Box::new(falseBranch))
        } else {
            None
        };
        Expr::If(Box::new(cond), Box::new(trueBranch), falseBranch)
    }

    fn parseFor(&mut self) -> Expr {
        self.expect(TokenKind::Keyword(KeywordKind::For));
        let pattern = self.parsePattern();
        self.expect(TokenKind::Keyword(KeywordKind::In));
        let source = self.parseExpr();
        self.expect(TokenKind::LeftBracket(BracketKind::Curly));
        self.undo();
        let body = self.parseExpr();
        Expr::For(pattern, Box::new(source), Box::new(body))
    }

    fn parseMatch(&mut self) -> Expr {
        self.expect(TokenKind::Keyword(KeywordKind::Match));
        let body = self.parseExpr();
        self.expect(TokenKind::LeftBracket(BracketKind::Curly));
        let mut branches = Vec::new();
        while !self.check(TokenKind::RightBracket(BracketKind::Curly)) {
            let pattern = self.parsePattern();
            self.expect(TokenKind::Arrow(ArrowKind::Right));
            let body = self.parseExpr();
            if self.check(TokenKind::Misc(MiscKind::Comma)) {
                self.expect(TokenKind::Misc(MiscKind::Comma));
            }
            branches.push(Branch { pattern, body });
        }
        self.expect(TokenKind::RightBracket(BracketKind::Curly));
        Expr::Match(Box::new(body), branches)
    }

    fn parseStatement(&mut self) -> (StatementKind, SemicolonRequirement) {
        match self.peek() {
            TokenKind::Keyword(KeywordKind::If) => {
                let expr = self.parseIf();
                (StatementKind::Expr(expr), SemicolonRequirement::Optional)
            }
            TokenKind::Keyword(KeywordKind::For) => {
                let expr = self.parseFor();
                (StatementKind::Expr(expr), SemicolonRequirement::Optional)
            }
            TokenKind::Keyword(KeywordKind::Match) => {
                let expr = self.parseMatch();
                (StatementKind::Expr(expr), SemicolonRequirement::Optional)
            }
            TokenKind::Keyword(KeywordKind::Let) => {
                self.expect(TokenKind::Keyword(KeywordKind::Let));
                let pattern = self.parsePattern();
                self.expect(TokenKind::Misc(MiscKind::Equal));
                let rhs = self.parseExpr();
                (
                    StatementKind::Let(pattern, rhs),
                    SemicolonRequirement::Required,
                )
            }
            _ => {
                let expr = self.parseExpr();
                if self.check(TokenKind::Misc(MiscKind::Equal)) {
                    self.expect(TokenKind::Misc(MiscKind::Equal));
                    let rhs = self.parseExpr();
                    (
                        StatementKind::Assign(expr, rhs),
                        SemicolonRequirement::Required,
                    )
                } else {
                    (
                        StatementKind::Expr(expr),
                        SemicolonRequirement::TrailingOptional,
                    )
                }
            }
        }
    }

    fn parseFieldAccessOrCall(&mut self) -> Expr {
        let mut current = self.parsePrimary();
        loop {
            if self.check(TokenKind::Misc(MiscKind::Dot)) {
                self.expect(TokenKind::Misc(MiscKind::Dot));
                let name = self.parseVarIdentifier();
                current = Expr::FieldAccess(Box::new(current), name);
            } else if self.check(TokenKind::LeftBracket(BracketKind::Paren)) {
                self.expect(TokenKind::LeftBracket(BracketKind::Paren));
                let mut args = Vec::new();
                while !self.check(TokenKind::RightBracket(BracketKind::Paren)) {
                    let arg = self.parseExpr();
                    args.push(arg);
                    if self.check(TokenKind::RightBracket(BracketKind::Paren)) {
                        break;
                    } else {
                        self.expect(TokenKind::Misc(MiscKind::Comma));
                    }
                }
                self.expect(TokenKind::RightBracket(BracketKind::Paren));
                current = Expr::Call(Box::new(current), args);
            } else {
                break;
            }
        }
        return current;
    }

    fn parseBinaryOp(&mut self, index: usize) -> Expr {
        let mut left = self.callNext(index);
        loop {
            let ops = &self.opTable[index];
            let mut matchingOp = None;
            for op in ops {
                if self.check(TokenKind::Op(*op)) {
                    matchingOp = Some(op.clone());
                    break;
                }
            }
            if let Some(op) = matchingOp {
                self.expect(TokenKind::Op(op));
                let rhs = self.callNext(index);
                let op = match op {
                    OperatorKind::And => BinaryOp::And,
                    OperatorKind::Or => BinaryOp::Or,
                    OperatorKind::Add => BinaryOp::Add,
                    OperatorKind::Sub => BinaryOp::Sub,
                    OperatorKind::Mul => BinaryOp::Mul,
                    OperatorKind::Div => BinaryOp::Div,
                    OperatorKind::Equal => BinaryOp::Equal,
                    OperatorKind::NotEqual => BinaryOp::NotEqual,
                    OperatorKind::LessThan => BinaryOp::LessThan,
                    OperatorKind::GreaterThan => BinaryOp::GreaterThan,
                    OperatorKind::LessThanOrEqual => BinaryOp::LessThanOrEqual,
                    OperatorKind::GreaterThanOrEqual => BinaryOp::GreaterThanOrEqual,
                };
                left = Expr::BinaryOp(op, Box::new(left), Box::new(rhs));
            } else {
                break;
            }
        }
        return left;
    }

    fn callNext(&mut self, index: usize) -> Expr {
        if index + 1 >= self.opTable.len() {
            self.parseFieldAccessOrCall()
        } else {
            self.parseBinaryOp(index + 1)
        }
    }

    fn parsePrimary(&mut self) -> Expr {
        match self.peek() {
            TokenKind::VarIdentifier => {
                let value = self.parseVarIdentifier();
                Expr::Value(value)
            }
            TokenKind::TypeIdentifier => {
                let value = self.parseTypeIdentifier();
                Expr::Name(value)
            }
            TokenKind::StringLiteral => {
                let tokenInfo = self.current().clone();
                self.step();
                if let Token::StringLiteral(value) = tokenInfo.token {
                    Expr::StringLiteral(value)
                } else {
                    unreachable!()
                }
            }
            TokenKind::IntegerLiteral => {
                let tokenInfo = self.current().clone();
                self.step();
                if let Token::IntegerLiteral(value) = tokenInfo.token {
                    Expr::IntegerLiteral(value)
                } else {
                    unreachable!()
                }
            }
            TokenKind::CharLiteral => {
                let tokenInfo = self.current().clone();
                self.step();
                if let Token::CharLiteral(value) = tokenInfo.token {
                    Expr::CharLiteral(value)
                } else {
                    unreachable!()
                }
            }
            TokenKind::Keyword(KeywordKind::ValueSelf) => {
                self.expect(TokenKind::Keyword(KeywordKind::ValueSelf));
                Expr::SelfValue
            }
            TokenKind::Keyword(KeywordKind::Return) => {
                self.expect(TokenKind::Keyword(KeywordKind::Return));
                let arg = if self.check(TokenKind::Misc(MiscKind::Semicolon))
                    || self.check(TokenKind::Misc(MiscKind::Comma))
                {
                    None
                } else {
                    Some(Box::new(self.parseExpr()))
                };
                Expr::Return(arg)
            }
            TokenKind::Keyword(KeywordKind::Continue) => {
                self.expect(TokenKind::Keyword(KeywordKind::Continue));
                let arg = if self.check(TokenKind::Misc(MiscKind::Semicolon))
                    || self.check(TokenKind::Misc(MiscKind::Comma))
                {
                    None
                } else {
                    Some(Box::new(self.parseExpr()))
                };
                Expr::Continue(arg)
            }
            TokenKind::Keyword(KeywordKind::Break) => {
                self.expect(TokenKind::Keyword(KeywordKind::Break));
                let arg = if self.check(TokenKind::Misc(MiscKind::Semicolon))
                    || self.check(TokenKind::Misc(MiscKind::Comma))
                {
                    None
                } else {
                    Some(Box::new(self.parseExpr()))
                };
                Expr::Break(arg)
            }
            TokenKind::Keyword(KeywordKind::If) => self.parseIf(),
            TokenKind::Keyword(KeywordKind::For) => self.parseFor(),
            TokenKind::Keyword(KeywordKind::Match) => self.parseMatch(),
            TokenKind::LeftBracket(BracketKind::Curly) => {
                let block = self.parseBlock();
                Expr::Block(block)
            }
            TokenKind::LeftBracket(BracketKind::Paren) => {
                self.expect(TokenKind::LeftBracket(BracketKind::Paren));
                let mut args = Vec::new();
                let mut commaAtEnd = false;
                while !self.check(TokenKind::RightBracket(BracketKind::Paren)) {
                    let arg = self.parseExpr();
                    args.push(arg);
                    if self.check(TokenKind::RightBracket(BracketKind::Paren)) {
                        commaAtEnd = false;
                        break;
                    } else {
                        self.expect(TokenKind::Misc(MiscKind::Comma));
                        commaAtEnd = true;
                    }
                }
                self.expect(TokenKind::RightBracket(BracketKind::Paren));
                if args.len() == 1 && !commaAtEnd {
                    args.remove(0)
                } else {
                    Expr::Tuple(args)
                }
            }
            kind => self.reportError2("<expr>", kind),
        }
    }

    fn parseExpr(&mut self) -> Expr {
        self.parseBinaryOp(0)
    }
}
